crates/sui-move-build/src/linters/native_primitives_only.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This analysis flags uses of the sui::coin::Coin struct in fields of other structs. In most cases
+//! it's preferable to use sui::balance::Balance instead to save space.
+
+use move_compiler::{
+    diag,
+    diagnostics::codes::{custom, DiagnosticInfo, Severity},
+    naming::ast as N,
+    shared::{CompilationEnv, Identifier},
+    typing::{ast as T, core::ProgramInfo, visitor::TypingVisitor},
+};
+use move_ir_types::location::Loc;
+use move_symbol_pool::Symbol;
+
+use super::{LinterDiagCategory, LINTER_DEFAULT_DIAG_CODE, LINT_WARNING_PREFIX};
+
+const NATIVE_PRIMITIVE_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagCategory::NativePrimitivesOnly as u8,
+    LINTER_DEFAULT_DIAG_CODE,
+    "passing non-native types to native functions",
+);
+
+pub struct NativePrimitivesOnlyVisitor;
+
+impl TypingVisitor for NativePrimitivesOnlyVisitor {
+    fn visit(
+        &mut self,
+        env: &mut CompilationEnv,
+        _program_info: &ProgramInfo,
+        program: &mut T::Program,
+    ) {
```

Comment:
You can replace this with `TypingVisitorConstructor` and `TypingVisitorContext` now that it has landed. It should reduce the amount of boilerplate you need to do